include/zephyr/bindesc.h

```diff
@@ -0,0 +1,369 @@
+ */
+#define BINDESC_TAG(type, id) ((BINDESC_TYPE_##type & 0xf) << 28 | (id))
+
+/*
+ * Corresponds to the definitions in scripts/west_commands/bindesc.py.
+ * Do not change without syncing the definitions in both files!
+ */
+
+/** The app version string such as "1.2.3" */
+#define BINDESC_TAG_APP_VERSION_STRING BINDESC_TAG(STR, 0x1000)
+
+/** The app version major such as 1 */
+#define BINDESC_TAG_APP_VERSION_MAJOR BINDESC_TAG(UINT, 0x1001)
+
+/** The app version minor such as 2 */
+#define BINDESC_TAG_APP_VERSION_MINOR BINDESC_TAG(UINT, 0x1002)
+
+/** The app version revision such as 3 */
+#define BINDESC_TAG_APP_VERSION_REVISION BINDESC_TAG(UINT, 0x1003)
+
+/** The kernel version string such as "3.4.0" */
+#define BINDESC_TAG_KERNEL_VERSION_STRING BINDESC_TAG(STR, 0x2000)
+
+/** The kernel version major such as 3 */
+#define BINDESC_TAG_KERNEL_VERSION_MAJOR BINDESC_TAG(UINT, 0x2001)
+
+/** The kernel version minor such as 4 */
+#define BINDESC_TAG_KERNEL_VERSION_MINOR BINDESC_TAG(UINT, 0x2002)
+
+/** The kernel version patchlevel such as 0 */
+#define BINDESC_TAG_KERNEL_VERSION_PATCHLEVEL BINDESC_TAG(UINT, 0x2003)
+
+/** The kernel version number such as 0x30400 */
+#define BINDESC_TAG_KERNEL_VERSION_NUMBER BINDESC_TAG(UINT, 0x2004)
+
+/** The year the image was compiled in */
+#define BINDESC_TAG_BUILD_TIME_YEAR BINDESC_TAG(UINT, 0x3000)
+
+/** The month of the year the image was compiled in */
+#define BINDESC_TAG_BUILD_TIME_MONTH BINDESC_TAG(UINT, 0x3001)
+
+/** The day of the month the image was compiled in */
+#define BINDESC_TAG_BUILD_TIME_DAY BINDESC_TAG(UINT, 0x3002)
+
+/** The hour of the day the image was compiled in */
+#define BINDESC_TAG_BUILD_TIME_HOUR BINDESC_TAG(UINT, 0x3003)
+
+/** The minute the image was compiled in */
+#define BINDESC_TAG_BUILD_TIME_MINUTE BINDESC_TAG(UINT, 0x3004)
+
+/** The second the image was compiled in */
+#define BINDESC_TAG_BUILD_TIME_SECOND BINDESC_TAG(UINT, 0x3005)
+
+/** The UNIX time (seconds since midnight of 1970/01/01) the image was compiled in */
+#define BINDESC_TAG_BUILD_TIME_UNIX BINDESC_TAG(UINT, 0x3006)
+
+/** The date and time of compilation such as "2023/02/05 00:07:04" */
+#define BINDESC_TAG_BUILD_DATE_TIME_STRING BINDESC_TAG(STR, 0x3007)
+
+/** The date of compilation such as "2023/02/05" */
+#define BINDESC_TAG_BUILD_DATE_STRING BINDESC_TAG(STR, 0x3008)
+
+/** The time of compilation such as "00:07:04" */
+#define BINDESC_TAG_BUILD_TIME_STRING BINDESC_TAG(STR, 0x3009)
+
+/** The name of the host that compiled the image */
+#define BINDESC_TAG_HOST_NAME BINDESC_TAG(STR, 0x4000)
+
+/** The C compiler name */
+#define BINDESC_TAG_C_COMPILER_NAME BINDESC_TAG(STR, 0x4001)
+
+/** The C compiler version */
+#define BINDESC_TAG_C_COMPILER_VERSION BINDESC_TAG(STR, 0x4002)
+
+/** The C++ compiler name */
+#define BINDESC_TAG_CXX_COMPILER_NAME BINDESC_TAG(STR, 0x4003)
+
+/** The C++ compiler version */
+#define BINDESC_TAG_CXX_COMPILER_VERSION BINDESC_TAG(STR, 0x4004)
```

Comment:
I started it from 0x1000 so that if a careless user accidentally uses 0/1/2/3... it won't cause any collision. The thing is that 32 bits is a huge field compared to what is stored in it. I really am thinking of shrinking the length (another overkill) and tag to 16 bits each. The only reason I haven't done this yet is that I fear this would cause issues with alignment or packing down the road. But I think with the proper flags and build assertions we can make it work and even save some space.

Reply:
Fair enough... explicit reservations are probably still a good idea

Reply:
Fair enough... explicit reservations are probably still a good idea?